firedrills.md

```diff
@@ -0,0 +1,782 @@
+
+Let’s illustrate this behavior by intentionally crashing CoreDNS pods:
+
+```bash
+node-1# EDITOR=nano kubectl -nkube-system edit configmaps/coredns
+<break config>
+```
+
+Restart the pods so they re-read the configuration:
+
+```bash
+node-1# kubectl -nkube-system delete pods -lk8s-app=kube-dns
+```
+
+Now let’s try to access a service from one of our debug pods:
+
+```bash
+node-1# kubectl exec -ti debug-jl8nr /bin/chroot /rootfs
+debug-pod# curl -k https://gravity-site.kube-system.svc.cluster.local:3009/healthz
+curl: (6) Could not resolve host: gravity-site.kube-system.svc.cluster.local
+```
+
+But CURLing the service IP directly works:
+
+```bash
+debug-pod# kctl get services | grep gravity-site
+debug-pod# curl -k https://10.100.135.21:3009/healthz
+{"info":"service is up and running","status":"ok"}
+```
+
+Thus far the issue seems to be laying on the DNS side. In order to further troubleshoot it, let’s also check the packets traffic using “tcpdump”. First, we need to find out the IP address of our container so we know what to filter by:
+
+```bash
+debug-pod# ifconfig eth0
+```
+
+Now, launch another terminal on node-1 and start sniffing:
+
+```bash
+node-1# sudo gravity exec tcpdump -n -i any host 10.244.4.21
+```
+
+Then, execute another “curl” from the debug container and see if “tcpdump” has captured any packets. Note, you may need to execute it a few times in order to get traffic on the correct node.
+
+```
+19:19:28.678062 IP 10.244.4.21.44651 > gravity.domain: 46573+ A? gravity-site.kube-system.svc.cluster.local. (60)
+19:19:28.678130 IP 10.244.4.21.44651 > gravity.domain: 47613+ AAAA? gravity-site.kube-system.svc.cluster.local. (60)
+19:19:28.678492 IP gravity.domain > 10.244.4.21.44651: 46573 ServFail 0/0/0 (60)
+```
+
+We can now see that “tcpdump” has captured the DNS request made by our pod and returned ServFail. This clearly points to the DNS issue so let’s check our CoreDNS pods and their logs:
+
+```bash
+node-1# kubectl -nkube-system get pods -owide
+node-1# kubectl -nkube-system logs coredns-86fsb
+```
+
+Let’s fix the configuration back and restart the pods:
+
+```bash
+node-1# EDITOR=nano kubectl -nkube-system edit configmaps/coredns
+<unbreak config>
+node-1# kubectl -nkube-system delete pods -lk8s-app=kube-dns
+```
+
+Keep in mind that DNS resolution issues can also be a side effect of broader networking issues, such as if there’s an issue with the overlay network we looked at before. In this case, you might get intermittent DNS resolution failures, e.g. when a pod attempts to resolve a service name and Kubernetes routes the DNS request to CoreDNS pod running on another node.
+
+### Scenario 1.5: Etcd
```

Comment:
A note should likely be added about master assignment, and this depends on the app. Some customers use a single master approach with multiple workers, which won't have the etcd redundancy (Same with replacing a failed node below).